### PR TITLE
lineup sol & image overhaul

### PR DESCRIPTION
**Pull request by SansPapyrus683**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

reeses puffs reeses puffs reeses puffs

### CODE REVIEW

File: solutions/usaco-965.mdx [hunk 1]
```diff
@@ -2,90 +2,272 @@
...
 ---
 
 [Official Analysis (C++)](http://www.usaco.org/current/data/sol_lineup_bronze_dec19.html)
 
-## Solution w/ Graphs
+## $\mathcal{O}(N)$ Solution With Graphs
 
-What if there were up to $10^5$ cows? Then neither of the solutions provided in
-the analysis would be fast enough.
+While the solutions provided in the editorial work just fine for the given bounds,
```
Comment (jessechoe10) [inline solutions/usaco-965.mdx:12]:
```suggestion
While the solutions provided in the editorial work for the given bounds,
```

File: solutions/usaco-965.mdx [hunk 2]
```diff
@@ -2,90 +2,272 @@
...
+there *is* a solution that works in just $\mathcal{O}(N)$!
```
Comment (jessechoe10) [inline solutions/usaco-965.mdx:13]:
```suggestion
there *is* a solution that works in just $\mathcal{O}(N)$, involving graphs!
```

File: solutions/usaco-965.mdx [hunk 3]
```diff
@@ -2,90 +2,272 @@
...
 
-Say that we draw an edge between two cows if they must be adjacent. Then if a
-solution exists, the graph must be a union of paths.
+Notice that since the input is guaranteed to be valid, we're always going to end
+up with "chains" of cows that we can arrange as we please. Using the sample given
```
Comment (jessechoe10) [inline solutions/usaco-965.mdx:16]:
```suggestion
up with multiple "chains" of cows that we can arrange. Using the sample given
```

File: solutions/usaco-965.mdx [hunk 4]
```diff
@@ -2,90 +2,272 @@
...
+in the problem, we'd get a "chain" representation like this:
```
Comment (jessechoe10) [inline solutions/usaco-965.mdx:17]:
```suggestion
in the problem, we'd get a "chain" representation as follows
```

File: solutions/usaco-965.mdx [hunk 5]
```diff
@@ -2,90 +2,272 @@
...
 
-(insert diagram)
+![](usaco-965/chains.png)
 
-We iterate over the cows in increasing lexicographical order. If a cow is not
-yet part of the answer and it is the endpoint of a path, then we add the entire
-path containing the cow to the answer.
+Note that cows that are not part of any chain can be considered their own chains
+of length 1 for implementation purposes.
```
Comment (jessechoe10) [inline solutions/usaco-965.mdx:22]: not needed

File: solutions/usaco-965.mdx [hunk 6]
```diff
@@ -2,90 +2,272 @@
...
+
+With this representation in mind, we can iterate through the cows in
```
Comment (jessechoe10) [inline solutions/usaco-965.mdx:24]:
```suggestion
With this graph representation in mind, we can iterate through the cows in
```

File: solutions/usaco-965.mdx [hunk 7]
```diff
@@ -2,90 +2,272 @@
...
+lexicographical order. When we come upon a cow that could be a possible
```
Comment (jessechoe10) [inline solutions/usaco-965.mdx:25]:
```suggestion
lexicographical order. When we visit a cow that could be a possible
```

File: solutions/usaco-965.mdx [hunk 8]
```diff
@@ -2,90 +2,272 @@
...
+start of a chain (AKA a cow that only has one required neighbor),
```
Comment (jessechoe10) [inline solutions/usaco-965.mdx:26]:
```suggestion
root of a chain (a cow that only has one neighbor),
```

File: solutions/usaco-965.mdx [hunk 9]
```diff
@@ -2,90 +2,272 @@
...
+<CPPSection>
 
 ```cpp
-#include <bits/stdc++.h>
-using namespace std;
+#include <fstream>
+#include <vector>
+#include <string>
+#include <algorithm>
+
```
Comment (jessechoe10) [inline solutions/usaco-965.mdx:41]: please use `using namespace std;`

File: solutions/usaco-965.mdx [hunk 10]
```diff
@@ -2,90 +2,272 @@
...
-		adj[c1].push_back(c2);
-		adj[c2].push_back(c1);
+	std::ifstream read("lineup.in");
+	int req_num;
+	read >> req_num;
+	vector<vector<int>> neighbors(COWS.size());
+	for (int r = 0; r < req_num; r++) {
+		string cow1;
+		string cow2;
+		string trash;
```
Comment (jessechoe10) [inline solutions/usaco-965.mdx:83]: better name pls

**Review by jessechoe10 — CHANGES_REQUESTED** [review]
also in ur explanations pls say connected components cuz that's the formal def

**Comment by SansPapyrus683** [comment]
by saying connected components, some information is lost
image
this triangle here would be a connected component, but it still wouldn't be a <i>chain</i>, really

File: solutions/usaco-965.mdx (hunk 10 above)
Comment (SansPapyrus683) [reply solutions/usaco-965.mdx:83]: i mean it is trash lmao

File: solutions/usaco-965.mdx (hunk 9 above)
Comment (SansPapyrus683) [reply solutions/usaco-965.mdx:41]: no

File: solutions/usaco-965.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply solutions/usaco-965.mdx:16]: chains is alr plural

File: solutions/usaco-965.mdx (hunk 4 above)
Comment (SansPapyrus683) [reply solutions/usaco-965.mdx:17]: not needed

File: solutions/usaco-965.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/usaco-965.mdx:22]: better explicit than implicit

File: solutions/usaco-965.mdx (hunk 6 above)
Comment (SansPapyrus683) [reply solutions/usaco-965.mdx:24]: why lol

File: solutions/usaco-965.mdx (hunk 8 above)
Comment (SansPapyrus683) [reply solutions/usaco-965.mdx:26]: why

File: solutions/usaco-965.mdx [hunk 11]
```diff
@@ -2,90 +2,272 @@
...
+		List<Integer>[] neighbors = new ArrayList[COWS.size()];
+		for (int c = 0; c < COWS.size(); c++) {
+			neighbors[c] = new ArrayList<>();
+		}
+		for (int r = 0; r < reqNum; r++) {
+			String[] words = read.readLine().split(" ");
+
+			// Convert the names to their index in the list
+			int cow1 = COWS.indexOf(words[0]);
+			int cow2 = COWS.indexOf(words[words.length - 1]);
```
Comment (nathangong) [inline solutions/usaco-965.mdx:168]: this makes the solution O(n^2)

Comment (SansPapyrus683) [reply solutions/usaco-965.mdx:168]: image
well, benq's solution does this as well
but ok i'll uh use a hashmap

Comment (SansPapyrus683) [reply solutions/usaco-965.mdx]: fixed

File: solutions/usaco-965.mdx (hunk 8 above)
Comment (jessechoe10) [reply solutions/usaco-965.mdx:26]: too informal

File: solutions/usaco-965.mdx (hunk 6 above)
Comment (jessechoe10) [reply solutions/usaco-965.mdx:24]: be more formal

File: solutions/usaco-965.mdx (hunk 5 above)
Comment (jessechoe10) [reply solutions/usaco-965.mdx:22]: Be more consise.

File: solutions/usaco-965.mdx (hunk 6 above)
Comment (SansPapyrus683) [reply solutions/usaco-965.mdx:24]: how does this make it more formal?

File: solutions/usaco-965.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/usaco-965.mdx:22]: > better explicit than implicit

File: solutions/usaco-965.mdx (hunk 8 above)
Comment (jessechoe10) [reply solutions/usaco-965.mdx:26]: don't use AKA

Comment (SansPapyrus683) [reply solutions/usaco-965.mdx:26]: i'll wait for third opinion

**Comment by SansPapyrus683** [comment]
no sotp

**Review by TruaShamu — APPROVED** [review]

File: solutions/usaco-965.mdx [hunk 12]
```diff
@@ -2,90 +2,266 @@
...
+				a, b = neighbors[at]
+				at, prev = b if a == prev else a, at
+
+			# Add the final element
+			added[at] = True
+			order.append(at)
+
+with open('lineup.out', 'w') as out:
+	for c in order:
+		print(COWS[c], file=out)
```
[the remark this answers is not in the document]
Comment (SansPapyrus683) [reply solutions/usaco-965.mdx:263]: wdym?

**Pull request merged**